examples/keygen.rs

```diff
@@ -0,0 +1,425 @@
+//! Demonstrates how to use Nova to produce a recursive proof of the correct execution of
+//! iterations of the key generation function.
+//! Generate public key using G1 curve, and prove using G2 curve.
+
+use bellperson::{
+  gadgets::{boolean::AllocatedBit, num::AllocatedNum},
+  ConstraintSystem, SynthesisError,
+};
+use ff::{Field, PrimeField, PrimeFieldBits};
+use generic_array::typenum::U5;
+use neptune::{
+  circuit::poseidon_hash,
+  poseidon::{Poseidon, PoseidonConstants},
+  Strength,
+};
+use nova_snark::{
+  gadgets::ecc::AllocatedPoint,
+  traits::{
+    circuit::{StepCircuit, TrivialTestCircuit},
+    Group as Nova_Group,
+  },
+  CompressedSNARK, PublicParams, RecursiveSNARK,
+};
+use pasta_curves::{
+  arithmetic::CurveAffine,
+  group::{Curve, Group},
+};
+use rand::{rngs::OsRng, RngCore};
+use subtle::Choice;
+
+type G1 = pasta_curves::pallas::Point;
+type G2 = pasta_curves::vesta::Point;
+type S1 = nova_snark::spartan_with_ipa_pc::RelaxedR1CSSNARK<G2>;
+type S2 = nova_snark::spartan_with_ipa_pc::RelaxedR1CSSNARK<G1>;
+
+#[derive(Debug, Clone, Copy)]
+pub struct SecretKey(pub <G1 as Group>::Scalar);
+
+impl SecretKey {
+  pub fn random(mut rng: impl RngCore) -> Self {
+    let secret = <G1 as Group>::Scalar::random(&mut rng);
+    Self(secret)
+  }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct PublicKey(pub G1);
+
+impl PublicKey {
+  pub fn from_secret_key(s: &SecretKey) -> Self {
+    let point = G1::generator() * s.0;
+    Self(point)
+  }
+}
+
+// An affine point coordinate that is on the curve.
+#[derive(Clone, Copy, Debug)]
+pub struct Coordinate<F>
+where
+  F: PrimeField<Repr = [u8; 32]>,
```

Comment:
Is it possible to avoid placing restrictions on Repr? I was able to drop such restrictions in the signature example, please see `examples/signature.rs`.